### PR TITLE
Added multistep methods as a sweeper

### PR DESCRIPTION
**Pull request by brownbaerchen**
As discussed in #327, you can abuse the sweeper class to do multistep methods.
I ended up doing this differently from what I initially thought and I don't know if this can be extended to arbitrary multistep methods.
In particular, at the moment it is required to use a fixed step size and restarts will probably also mess everything up.

I thought I start with a simple implementation of the most basic things and anybody is welcome to improve on this.
@lisawim, maybe we can discuss how we get the methods you want into this notation? I don't know exactly what you did with multistep methods so far.

I implemented methods with at most 2 steps, so not overly exciting, but for this I didn't have any issues with generating starting values, as they can be one order below the scheme.

I test the order of accuracy in time, the stability and the number of right hand side evaluations.

### CODE REVIEW

**Comment by pancetta** [comment]
So, this is multi-step methods within a single time-step, yes? I'd not call it multi-step methods then,  because people would expect something else under this name.

**Comment by brownbaerchen** [comment]
> So, this is multi-step methods within a single time-step, yes? I'd not call it multi-step methods then, because people would expect something else under this name.

No, it's multistep methods. It stores the solutions from past steps and uses these to compute a solution to the current step. At least in the book I copied these methods from they referred to these methods as multistep. Of course every but the two step Adams-Moulton method are very boring examples of multistep methods, but I believe taxonomy allows to call single-stage one-step methods either one-step or multistep.

File: pySDC/implementations/sweeper_classes/Multistep.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+    def __init__(self, params):
+        alpha = [-1.0]
+        beta = [1.0, 0.0]
+        super().__init__(params, alpha, beta)
+
+
+class BackwardEuler(MultiStep):
+    """
+    Do you like mess? Me neither! Let me lure you in!
```
Comment (pancetta) [inline pySDC/implementations/sweeper_classes/Multistep.py:177]: Yeah.. nice try.

**Comment by pancetta** [comment]
Ok, ok, I see. Not sure I get it, yet: You store the solution of how many previous steps and how many previous nodes? Is it: last step, `len(alpha)` nodes? Or only the last node of the `len(alpha)` steps?

**Comment by brownbaerchen** [comment]
> Ok, ok, I see. Not sure I get it, yet: You store the solution of how many previous steps and how many previous nodes? Is it: last step, `len(alpha)` nodes? Or only the last node of the `len(alpha)` steps?

I only store the solutions to previous steps, not any of the nodes. That would also be interesting, but this PR is just about getting more established methods for comparisons into pySDC. Lisa expressed some interest in using this for her DAE project, so I went ahead and did this simple stuff.
Actually, there should not be any particular nodes. Just one for initial conditions and one for the solution of the step. But I did forget to fix this...

As always, I don't know anything about multistep methods, but I followed my favourite book, the Leveque finite difference book, and that had the convention: N alpha values for the previous solutions and N+1 beta values for the right hand side evaluations at the previous solutions and at the current one.

I should maybe elaborate on the documentation...

So I have two todos now:
- [ ] More documentation on alpha and beta parameters
- [ ] Fix the "collocation" to single node

Anything else?

File: pySDC/implementations/sweeper_classes/Multistep.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
+from pySDC.core.Sweeper import sweeper, _Pars
+
+
+class Cache(object):
```
Comment (lisawim) [inline pySDC/implementations/sweeper_classes/Multistep.py:4]: That's cool! I like the class, pretty good idea! So you wrote a class having the values from previous steps and can update them using the `update` method? I was thinking way too complicated when I was implementing that..

Comment (brownbaerchen) [reply pySDC/implementations/sweeper_classes/Multistep.py:4]: Yes, but this is actually a bit too simple. I know both of us want to restart time steps and that is not really supported with this simple cache. But we can add that later. For instance, we can keep one extra solution in the cache and do a reverse `update` if we restart a step. 
We should sit down while we're here at Hamburg at some point so you can show me what you already did and we can merge our efforts.

**Comment by pancetta** [comment]
So, what's the status of this?

**Comment by lisawim** [comment]
After some brainstorming with @brownbaerchen in Hamburg starting values for the LM methods are not always available and in case of using more than one previous time step to start with, starting values has to be generated by another method, e.g. RK methods. This seems to be not implemented yet, but I'm not sure if he wants to add this here. Please @brownbaerchen correct me if I'm wrong.

**Comment by brownbaerchen** [comment]
Of course @tlunet's input is always appreciated! But don't feel obligated if you don't have the time.

I intended this is as a starting point if someone ever really wants to use LMMs. So far, the methods are of limited use. I guess the third order implicit two step method is not too bad, as long as you have fixed step size and no restarts, which has been the only way of running the code for a long time. #331 made it easy to access Butcher tableaux outside of RK sweepers. Now, all that needs to be done to add LMMs of order larger than 3 is to basically copy-paste the `update_nodes` function from the RK sweeper and plug in the Q matrices from the RK methods. I did not plan on doing this currently, because I am not sure that anybody will be using it.

Of course, @lisawim wants to use LMMs, but I was unable to get high order in time with the simple DAE problem from @kyrillh without putting any sort of thought into this. So I guess it makes more sense to move forward with this when we know how to do DAEs with RK.
@lisawim and I did almost the same thing apart from her doing DAEs and me ODEs. However, I tested that the LMMs do not perform unnecessary right hand side evaluations. If it's fine with @lisawim, we could merge this and you could inherit from this, only changing the `update_nodes` function to DAE stuff and implementing the coefficients?

**Comment by lisawim** [comment]
That's a good idea, I'm completely fine with that! I'll also have to see what's going on with the order of accuracy..

**Pull request merged**